cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java

```diff
@@ -128,55 +132,70 @@
   }
 
   @Override
-  public synchronized RuntimeInfo lookup(final RunId runId) {
-    RuntimeInfo runtimeInfo = super.lookup(runId);
-    if (runtimeInfo != null) {
-      return runtimeInfo;
+  protected RuntimeInfo createRuntimeInfo(ProgramController controller, Program program) {
+    if (controller instanceof AbstractTwillProgramController) {
+      RunId twillRunId = ((AbstractTwillProgramController) controller).getTwillRunId();
+      return new SimpleRuntimeInfo(controller, program, twillRunId);
     }
+    return null;
+  }
 
-    // Lookup all live applications and look for the one that matches runId
-    String appName = null;
-    TwillController controller = null;
-    for (TwillRunner.LiveInfo liveInfo : twillRunner.lookupLive()) {
-      for (TwillController c : liveInfo.getControllers()) {
-        if (c.getRunId().equals(runId)) {
-          appName = liveInfo.getApplicationName();
-          controller = c;
-          break;
-        }
-      }
-      if (controller != null) {
-        break;
+  private synchronized boolean isTwillRunIdCached(RunId twillRunId) {
+    for (RuntimeInfo runtimeInfo : getRuntimeInfos()) {
+      if (twillRunId.equals(runtimeInfo.getTwillRunId())) {
+        return true;
       }
     }
+    return false;
+  }
 
-    if (controller == null) {
-      LOG.info("No running instance found for RunId {}", runId);
-      // TODO (ENG-2623): How about mapreduce job?
-      return null;
+  @Override
+  public synchronized RuntimeInfo lookup(Id.Program programId, final RunId runId) {
+    RuntimeInfo runtimeInfo = super.lookup(programId, runId);
+    if (runtimeInfo != null) {
+      return runtimeInfo;
     }
 
-    Matcher matcher = APP_NAME_PATTERN.matcher(appName);
-    if (!matcher.matches()) {
-      LOG.warn("Unrecognized application name pattern {}", appName);
+    RunRecord record = store.getRun(programId, runId.getId());
```

Comment:
Remember that hitting Store has much higher cost that looping the twill live list. Hitting Store is transactionally and involves quite a number of RPC round trip and scan on HBase. Looping twill live list is all happening in-memory.